### PR TITLE
Game tests + sha256 builtin

### PR DESCRIPTION
**Pull request by j-mueller**
Adds everything that we need for the demo

* sha2 and sha3 primitives
* end-to-end tests for game and crowdfunding contract
* better comments for the crowdfunding contract

### CODE REVIEW

**Comment by michaelpj** [comment]
You need to regenerate the package set.

File: plutus-tx-plugin/test/Plugin/Spec.hs [hunk 1]
```diff
@@ -82,6 +82,8 @@ primitives = testNested "primitives" [
   --, goldenPlc "blocknum" blocknumPlc
   , goldenPir "bytestring" bytestring
   , goldenEval "bytestringApply" [ getPlc bytestring, unsafeLiftProgram ("hello"::ByteString) ]
+  , goldenEval "sha2_256" [ getPlc sha2, unsafeLiftProgram ("hello" :: ByteString)]
```
Comment (j-mueller) [inline plutus-tx-plugin/test/Plugin/Spec.hs]: @effectfully @michaelpj Could you have a look at this test please? It results in `Failure` (see below), I don't know why!

Comment (effectfully) [reply plutus-tx-plugin/test/Plugin/Spec.hs]: Could you please show the output of `sha2 "hello"`?

Comment (effectfully) [reply plutus-tx-plugin/test/Plugin/Spec.hs]: It seems like the output somehow does not fit into a `ByteString` of length 32.

Comment (effectfully) [reply plutus-tx-plugin/test/Plugin/Spec.hs]: Or at least I can't think of a different reason a `Failure` can be returned.

Comment (j-mueller) [reply plutus-tx-plugin/test/Plugin/Spec.hs]: yeah looks like that is the reason. The hash has 63 bytes.

```
>Hash.sha2 ("hello" :: BSL.ByteString)
"\159\CAN,\CAN\242\CANM\CAN\186\CAN_\CAN\176\CAN\163\SO\CAN&\CAN\232\CAN;\CAN*\CAN\197\CAN\185\CAN\226\CAN\158\CAN\ESC\SYN\CAN\RS\CAN\\\CAN\US\CAN\167\CANB\CAN^\CANs\EOT\CAN3\CANb\CAN\147\CAN\139\CAN\152\CAN$\255"
>BSL.length (Hash.sha2 ("hello" :: BSL.ByteString))
63
>
```

Comment (j-mueller) [reply plutus-tx-plugin/test/Plugin/Spec.hs]: ok. Looks like the size in the spec is wrong. I'll get back to this later today.

Comment (j-mueller) [reply plutus-tx-plugin/test/Plugin/Spec.hs]: or rather, my implementation of sha2 and sha3

**Comment by j-mueller** [comment]
I'd like to merge this soon, looks like the builtins are working. 

@vmchale Are you OK with the new module?

**Pull request merged**